collector/src/artifact_stats.rs

```diff
@@ -102,16 +102,41 @@ impl ArtifactStats {
     }
 }
 
-/// Tries to match hashes produces by rustc in mangled symbol names.
-static RUSTC_HASH_REGEX: OnceLock<Regex> = OnceLock::new();
-
 /// Demangle the symbol and remove rustc mangling hashes.
+///
+/// Normalizes the following things, in the following order:
+/// - Demangles the symbol.
+/// - Removes `.cold` and `.warm` from the end of the symbol, to merge cold and hot parts of a function
+/// into the same symbol.
+/// - Removes rustc hashes from the symbol, e.g. `foo::[abcdef]` -> `foo::[]` or
+/// `foo::abcd` -> `foo`.
```

Comment:
I'm not sure I understand this  `foo::abcd` -> `foo` transform. What does `normalize_short_hash` look like without it?

Reply:
Yeah, this is tricky, because `abcd` can be a normal name, or a hash. The regex tries to remove hashes, but keep normal names. So `std::fs::File` is obviously ok, but `std::fs::File::abcde231312fa` is not. But what is and isn't a hash isn't really trivial to figure out.

Reply:
Then maybe we should punt on this specific normalization step for later, if it's hard to avoid such false positives.

Reply:
I wouldn't say that it has caused issues so far. Right now the regex is `::[a-z0-9]{15,17}$`. Having a such a long identifier (without any underscore or hyphen) at the end of the demangled name is probably very uncommon.

Reply:
I don't know if they always appear in specific constructs, e.g. for closures or something. If that's the case maybe we should normalize the whole and not the parts. 

Writing this also makes me realize that eg ".cold" and ".warm" are handled differently than the numeric suffix in ".llvm.2132123131". But really all these are the same should likely be normalized on the mangled symbols as vendor specific suffixes after a dot.

Reply:
I remember that there were some weird things with the dot, and just removing everything after a dot didn't work I think, but I don't remember the specifics :( Indeed it is weird that the numeric suffix is removed just by demangling, but not the `.cold` annotations.

Reply:
Agreed. It could be a legacy/v0 thing, or a missing feature in the demangler.